File: packages/go-agentenkit/ports/queue.go
package ports

import (
	"context"
	"errors"
	"time"
)

// JobKind says why a job was enqueued. A fresh user run is the default; the
// platform's own follow-ups name themselves so a queue can log them, order
// them and cap only the kind that brings new work in.
type JobKind string

const (
	// JobDispatch is a fresh run from Run: the only kind a depth cap refuses.
	JobDispatch JobKind = ""
	// JobRetry is the same run trying again after a failure (§2.8).
	JobRetry JobKind = "retry"
	// JobRedrive is a job that found the run lock held by an older run.
	JobRedrive JobKind = "redrive"
	// JobResume continues a parked run after an answer (§2.5).
	JobResume JobKind = "resume"
	// JobExpiry is a park's own deadline (§2.5).
	JobExpiry JobKind = "expiry"
	// JobReclaim re-dispatches an orphaned thread (§2.5).
	JobReclaim JobKind = "reclaim"
)

// PriorityLow is the priority of the platform's own housekeeping: park
// expiries, reclaims and lock redrives. A user's message is never queued
// behind them.
const PriorityLow = -10

var (
	// ErrQueueFull is what Enqueue returns when a depth cap is set and the
	// queue has reached it. Only a fresh dispatch is ever refused; retries,
	// redrives, resumes and expiries always go in, or a run already under way
	// would be stranded.
	ErrQueueFull = errors.New("agentenkit: queue is full")
	// ErrPayloadTooLarge is a job whose encoded payload is over the adapter's
	// cap.
	ErrPayloadTooLarge = errors.New("agentenkit: job payload too large")
	// ErrDuplicateJob is an enqueue whose Key is already waiting or running.
	ErrDuplicateJob = errors.New("agentenkit: a job with this key is already queued")
	// ErrUnsupported is an adapter's answer to an operation it cannot do,
	// such as counting on a queue with no read side.
	ErrUnsupported = errors.New("agentenkit: not supported by this adapter")
)

// EnqueueOptions tunes one dispatch.
type EnqueueOptions struct {
	// Delay holds the job before delivering it: the HITL expiry (§2.5), a
	// job blocked by an older run's lock (§2.8) and a failed run's retry
	// backoff.
	//
	// An adapter that cannot delay may deliver immediately (every caller
	// treats an early arrival as a no-op) but it must NOT fail, or a park
	// would fail the run that scheduled it.
	Delay time.Duration
	// Key dedupes. At most one live row per key: a second enqueue with the
	// same key is refused with ErrDuplicateJob, and Cancel drops the waiting
	// row for a key. Empty means no dedupe.
	Key string
	// Priority orders ready rows: higher first, then oldest first. Zero is a
	// normal dispatch; the platform's housekeeping uses PriorityLow.
	Priority int
}

// QueueStats is what a queue can say about itself. Every count is scoped to
// the consumer's own namespace.
type QueueStats struct {
	// Ready is jobs a worker could take right now.
	Ready int `json:"ready"`
	// Delayed is jobs held for a future run time: park expiries, retry
	// backoffs. Not backlog.
	Delayed int `json:"delayed"`
	// InFlight is jobs a worker holds a lease on.
	InFlight int `json:"inFlight"`
	// Dead is jobs the queue gave up on and kept for an operator.
	Dead int `json:"dead"`
	// OldestReadyMs is how long the oldest ready job has been waiting. The
	// number that says a queue is stuck rather than merely busy.
	OldestReadyMs int64 `json:"oldestReadyMs"`
	// Paused is true while the consumer is told not to claim.
	Paused bool `json:"paused"`
	// LastClaimAt is when this process last took a job; nil when it never has.
	LastClaimAt *time.Time `json:"lastClaimAt,omitempty"`
	// ClaimErrors is how many claims in a row have failed in this process.
	ClaimErrors int `json:"claimErrors"`
}

// QueuedJob is one row as the queue sees it, for a host that asks where a
// run's job is.
type QueuedJob struct {
	ID          string     `json:"id"`
	RunID       string     `json:"runId"`
	ThreadID    string     `json:"threadId"`
	Kind        JobKind    `json:"kind"`
	Attempts    int        `json:"attempts"`
	RunAt       time.Time  `json:"runAt"`
	LockedUntil *time.Time `json:"lockedUntil,omitempty"`
	// Position is how many ready jobs are ahead of it; -1 when the adapter
	// cannot say.
	Position int `json:"position"`
}

// Queue is the durable dispatch port (§2.8). Delivery is at-least-once; the
// engine's state guard and the run lock make double dispatch a no-op.
//
// Beside Enqueue the port carries the little an engine and a host need to
// see the queue: a way to withdraw a job by key, a way to find a run's job,
// and counts. An adapter that has no read side answers ErrUnsupported to
// Find and Stats; callers treat that as "unknown", never as a failure.
type Queue interface {
	Enqueue(ctx context.Context, job RunJob, opts *EnqueueOptions) error
	// Cancel drops every waiting job enqueued under key. A key nothing is
	// waiting under is not an error. Best-effort by contract: a delivered
	// row is a correct no-op anyway.
	Cancel(ctx context.Context, key string) error
	// Find returns the earliest waiting or running job for a run, or nil
	// when the queue holds none.
	Find(ctx context.Context, runID string) (*QueuedJob, error)
	// Stats counts what waits, runs and died.
	Stats(ctx context.Context) (QueueStats, error)
}

File: packages/agentenkit/src/ports/queue.ts
import type { RunJob } from '../core/types.js';

/** The priority of the platform's own housekeeping: park expiries, reclaims
 *  and lock redrives. A user's message is never queued behind them. */
export const PRIORITY_LOW = -10;

/** `enqueue` refused a fresh dispatch: a depth cap is set and the queue has
 *  reached it. Only a fresh dispatch is ever refused; retries, redrives,
 *  resumes and expiries always go in, or a run already under way would be
 *  stranded. */
export class QueueFullError extends Error {
  constructor(message = 'queue is full') {
    super(message);
    this.name = 'QueueFullError';
  }
}

/** A job whose encoded payload is over the adapter's cap. */
export class PayloadTooLargeError extends Error {
  constructor(message = 'job payload too large') {
    super(message);
    this.name = 'PayloadTooLargeError';
  }
}

/** An enqueue whose `key` is already waiting or running. */
export class DuplicateJobError extends Error {
  constructor(key: string) {
    super(`a job with this key is already queued: ${key}`);
    this.name = 'DuplicateJobError';
  }
}

/** An adapter's answer to an operation it cannot do, such as counting on a
 *  queue with no read side. Callers treat it as "unknown", never as a
 *  failure. */
export class UnsupportedError extends Error {
  constructor(what: string) {
    super(`not supported by this adapter: ${what}`);
    this.name = 'UnsupportedError';
  }
}

export interface EnqueueOptions {
  /** Hold the job for this many seconds before delivering it: the HITL expiry
   *  (§2.5), a job blocked by an older run's lock (§2.8) and a failed run's
   *  retry backoff.
   *
   *  An adapter that cannot delay may deliver immediately — every caller
   *  treats an early arrival as a no-op — but it must NOT throw, or a park
   *  would fail the run that scheduled it. Note the shape of the reference
   *  adapter: QStash supports delays on publish only and rejects them on
   *  queue enqueue, so QStashQueue publishes delayed jobs and queues the
   *  rest. */
  delaySeconds?: number;
  /** Dedupes. At most one live job per key: a second enqueue with the same
   *  key is refused with `DuplicateJobError`, and `cancel` drops the waiting
   *  job for a key. Omitted means no dedupe. */
  key?: string;
  /** Orders ready jobs: higher first, then oldest first. 0 is a normal
   *  dispatch; the platform's housekeeping uses `PRIORITY_LOW`. */
  priority?: number;
}

/** What a queue can say about itself. */
export interface QueueStats {
  /** Jobs a worker could take right now. */
  ready: number;
  /** Jobs held for a future run time: park expiries, retry backoffs. Not
   *  backlog. */
  delayed: number;
  /** Jobs a worker holds. */
  inFlight: number;
  /** Jobs the queue gave up on and kept for an operator. */
  dead: number;
  /** How long the oldest ready job has been waiting: the number that says a
   *  queue is stuck rather than merely busy. */
  oldestReadyMs: number;
  /** True while the consumer is told not to claim. */
  paused: boolean;
  /** When this process last took a job. */
  lastClaimAt?: Date;
  /** How many claims in a row have failed in this process. */
  claimErrors: number;
}

/** One job as the queue sees it, for a host that asks where a run's job is. */
export interface QueuedJob {
  id: string;
  runId: string;
  threadId: string;
  kind: RunJob['kind'];
  attempts: number;
  runAt: Date;
  lockedUntil?: Date;
  /** How many ready jobs are ahead of it; -1 when the adapter cannot say. */
  position: number;
}

/** Durable dispatch port (§2.8). Delivery is at-least-once — the engine's
 *  state guard and the run lock make double dispatch a no-op.
 *
 *  Beside `enqueue` the port carries the little an engine and a host need to
 *  see the queue: a way to withdraw a job by key, a way to find a run's job,
 *  and counts. An adapter with no read side throws `UnsupportedError` from
 *  `find` and `stats`; callers treat that as "unknown", never as a failure. */
export interface Queue {
  enqueue(job: RunJob, opts?: EnqueueOptions): Promise<void>;
  /** Drop every waiting job enqueued under `key`. A key nothing is waiting
   *  under is not an error. Best-effort by contract: a delivered job is a
   *  correct no-op anyway. */
  cancel(key: string): Promise<void>;
  /** The earliest waiting or running job for a run, or null when the queue
   *  holds none. */
  find(runId: string): Promise<QueuedJob | null>;
  /** Count what waits, runs and died. */
  stats(): Promise<QueueStats>;
}
